students/Harry_Maher/session06/mailroom4.py

```diff
@@ -0,0 +1,123 @@
+        Include Donor Name, total donated, number of donations and average donation amount as values in each row. You do not need to print out all their donations, just the summary info.
+        Using string formatting, format the output rows as nicely as possible. The end result should be tabular (values in each column should align with those above and below)
+        After printing this report, return to the original prompt.
+    At any point, the user should be able to quit their current task and return to the original prompt.
+    From the original prompt, the user should be able to quit the script cleanly.
+
+Your report should look something like this:
+
+Donor Name                | Total Given | Num Gifts | Average Gift
+------------------------------------------------------------------
+William Gates, III         $  653784.49           2  $   326892.24
+Mark Zuckerberg            $   16396.10           3  $     5465.37
+Jeff Bezos                 $     877.33           1  $      877.33
+Paul Allen                 $     708.42           3  $      236.14
+
+
+
+'''
+import os
+
+# Just gonna make these global. Lazy and waiting until we *have to* implement classes...
+donor_dict = {"Bill":[2.75,3.12], "Bob": [52.75,15.11], \
+"Jim": [27.36,2.07] , "Ann": [12.76,1.01], "Beth": [31245.75]}
+donors = [donor for donor in donor_dict.keys()] 
+
+
+def pick_ty_recipient(name = None):
+    # Annoyingly refactored for testing. Sorry.    
+    if name == None: 
+        name = input("Pick or add a name (or type list)!\n")
+    if name.strip().lower() == "list":
+        print("\n".join(donors))
+        # If some fool wants to keep typing "list" for some reason:
+        name = pick_ty_recipient() 
+    return name.title().strip()
+
+
+def thank_you_txt(name):
+        return f"Dear {name},\n\n \tThanks so much for your generous\
+ donations totaling ${sum(donor_dict[name]):.2f}.\n \tWe are\
+ all incredibly grateful because blah blah. \n\n-NGO"
+
+
+def send_ty(name = None):
+    """Prints out a thank you letter to specified person, or to new person."""
+    name = pick_ty_recipient(name)
+    if name not in donors and (name !="list"):
+        donor_dict[name] = [0,0]
+        donation = input("How much did they donate?\n")
+        try:
+            donor_dict[name].append(float(donation))
+        except ValueError:
+            donation = input("Donation needs to be a number")
+        donor_dict[name].append(float(donation))
+    print(thank_you_txt(name))
+    return thank_you_txt(name)
+
+
+def generate_report():
+    """print a report of donation averages by donor."""
+    report = "{:16}|{:^13}|{:^11}|{:>13}\n".format("Donor Name", "Total Given", "Num Gifts", "Average Gift")
+    for name in donors:
+        report+="{:17}${:>12.2f} {:>11} ${:>12.2f}\n".format(name, sum(donor_dict[name]), len(donor_dict[name]),(sum(donor_dict[name])/len(donor_dict[name])))
```

Comment:
You could assign sum(donor_dict[name]) and len(donor_dict[name]) to variable names so that you only need to calculate this once. That will make your function more efficient.